### ISSUE TITLE
about MLM softlabel implementation

### ISSUE THREAD

**Issue by zhezh** [salesforce/ALBEF#58]
```python
soft_labels=F.softmax(logits_m, dim=-1)
```

```python
if soft_labels is not None:
	  loss_distill = -torch.sum(F.log_softmax(prediction_scores, dim=1) * soft_labels, dim=-1)
	  loss_distill = loss_distill[labels != -100].mean()
	  masked_lm_loss = (1 - alpha) * masked_lm_loss + alpha * loss_distill
```

The `logits_m` and `prediction_scores` are both of shape `(batch, seq_len, n_vocab)`.

My question is why softmax on dim=1 for `prediction_scores` , but not on dim=-1 ?

**Comment by LiJunnan1992** [comment]
It is a typo and has been fixed. Thanks for pointing it out! We will run experiments to make sure that results are not affected.

**Comment by sanyalsunny111** [comment]
`if soft_labels is not None:
	  loss_distill = -torch.sum(F.log_softmax(prediction_scores, dim=1) * soft_labels, dim=-1)
	  loss_distill = loss_distill[labels != -100].mean()
	  masked_lm_loss = (1 - alpha) * masked_lm_loss + alpha * loss_distill`
	  
	  My question is based on equation 7 in ALBEF paper you guys have claimed to be using KL divergence but I don't see you guys have used KL-divergence in your loss function? Could you please explain?

**Comment by LiJunnan1992** [comment]
> `if soft_labels is not None: loss_distill = -torch.sum(F.log_softmax(prediction_scores, dim=1) * soft_labels, dim=-1) loss_distill = loss_distill[labels != -100].mean() masked_lm_loss = (1 - alpha) * masked_lm_loss + alpha * loss_distill`
> 
> ### My question is based on equation 7 in ALBEF paper you guys have claimed to be using KL divergence but I don't see you guys have used KL-divergence in your loss function? Could you please explain?

KL-Div is equivalent to cross-entropy loss when the target distribution does not have gradient.

**Comment by ChenYutongTHU** [comment]
Hello~ I notice that when computing mlm loss,  you were doing next-token prediction. To do so, you shifted the target label left by one step in [models/xbert.py](https://github.com/salesforce/ALBEF/blob/main/models/xbert.py).
```
        if labels is not None:
            # we are doing next-token prediction; shift prediction scores and input ids by one
            shifted_prediction_scores = prediction_scores[:, :-1, :].contiguous()
            labels = labels[:, 1:].contiguous()
            loss_fct = CrossEntropyLoss(reduction=reduction)
            lm_loss = loss_fct(shifted_prediction_scores.view(-1, self.config.vocab_size), labels.view(-1))
            lm_loss = lm_loss.view(prediction_scores.size(0),-1).sum(1)
```
However, you set target for unmasked tokens to -100 to escape loss computation ([models/model_pretrain.py](https://github.com/salesforce/ALBEF/blob/main/models/model_pretrain.py)).
```
        if targets is not None:
            targets[~masked_indices] = -100 # We only compute loss on masked tokens 
```
If the right neighbor of a masked token is unmasked (very likely), the shifted target for the masked token is -100 and thus the MLM training for this token is invalid.  I guess it's a bug here.

Also, could you please explain why you used next-prediction strategy here  (more GPT-like). It makes more sense to me to recover the masked token when you load a pretrained BERT model. And the idea of predicting the token right next to the masked token is strange ..
